### PR TITLE
feat(rrb): Support for running a pipeline between scale up and disable

### PR DESCRIPTION
**Pull request by ajordens**
Coordinates for both the new and old server groups are passed along as
pipeline parameters.

image

### CODE REVIEW

**Comment by ajordens** [comment]
Before I go too much farther down this path, how does this look as an approach @tomaslin / @lwander?

Chatted with a couple internal stakeholders and I'll probably expose some configuration that would allow for the validation pipeline to fall through to a manual judgment if it failed.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy [hunk 1]
```diff
@@ -86,72 +90,118 @@ class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
...
+        resizeServerGroupStage.type,
+        "Grow to $p% Desired Size",
+        resizeContext,
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
       )
+      stages << resizeStage
+
+      // an expression to grab newly deployed server group at runtime (ie. the server group being resized up)
+      def deployedServerGroupName = '${' + "#stage('${resizeStage.id}')['context']['asgName']" + '}'.toString()
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy]: This seemed like a handy thing to be able to do ... required making `#stage` capable of working off an id vs name.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy [hunk 2]
```diff
@@ -86,72 +90,118 @@ class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
...
   }
 
+  List<Stage> getBeforeCleanupStages(Stage parentStage,
+                                     RollingRedBlackStageData stageData,
+                                     ResizeStrategy.Source source,
+                                     String deployedServerGroupName,
+                                     int percentageComplete) {
+    def stages = []
+
+    if (stageData.getDelayBeforeCleanup()) {
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy]: Not likely you would have both a delay + pipeline, I plan on preventing it in the UI.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy [hunk 3]
```diff
@@ -86,72 +90,118 @@ class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
...
+        serverGroupName: source.serverGroupName,
+        account        : source.credentials,
+        cloudProvider  : source.cloudProvider
+      ]
+
+      def pipelineContext = [
+        pipelineApplication: stageData.pipelineBeforeCleanup.application,
+        pipelineId         : stageData.pipelineBeforeCleanup.pipelineId,
+        pipelineParameters : [
+          "deployedServerGroup": serverGroupCoordinates + [
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy]: Seemed like reasonable things to pass down to the validation pipeline.

Internally we're not quite sure what these pipelines would do right now, but they're an easy extension point. Once usage picks up, common pattern might be extracted into new stages (ie. something that checks our internal metric system for deviation).

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/support/StageData.groovy [hunk 4]
```diff
@@ -58,6 +63,10 @@ class StageData {
     return account ?: credentials
   }
 
+  long getDelayBeforeCleanup() {
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/support/StageData.groovy]: Not tied to names but wanted something reasonably consistent and we've got a few eager souls that have already started using `delayBeforeDisableSec`.

**Comment by ajordens** [comment]
image

**Comment by lwander** [comment]
This looks really neat

**Pull request merged**